README.md
# SYNOPSIS 
[![NPM Package](https://img.shields.io/npm/v/vaporyjs-vm.svg?style=flat-square)](https://www.npmjs.org/package/vaporyjs-vm)
[![Build Status](https://img.shields.io/travis/vaporyjs/vaporyjs-vm.svg?branch=master&style=flat-square)](https://travis-ci.org/vaporyjs/vaporyjs-vm)
[![Gitter](https://img.shields.io/gitter/room/vapory/vaporyjs-lib.svg?style=flat-square)](https://gitter.im/vapory/vaporyjs-lib) or #vaporyjs on freenode  

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard) 

Implements Vapory's VM in JS

# INSTALL
`npm install vaporyjs-vm`

# USAGE
```javascript
var VM = require('vaporyjs-vm')

//create a new VM instance
var vm = new VM()
var code = '7f4e616d65526567000000000000000000000000000000000000000000000000003055307f4e616d6552656700000000000000000000000000000000000000000000000000557f436f6e666967000000000000000000000000000000000000000000000000000073661005d2720d855f1d9976f88bb10c1a3398c77f5573661005d2720d855f1d9976f88bb10c1a3398c77f7f436f6e6669670000000000000000000000000000000000000000000000000000553360455560df806100c56000396000f3007f726567697374657200000000000000000000000000000000000000000000000060003514156053576020355415603257005b335415603e5760003354555b6020353360006000a233602035556020353355005b60007f756e72656769737465720000000000000000000000000000000000000000000060003514156082575033545b1560995733335460006000a2600033545560003355005b60007f6b696c6c00000000000000000000000000000000000000000000000000000000600035141560cb575060455433145b1560d25733ff5b6000355460005260206000f3'
//code needs to be a buffer
code = Buffer.from(code, 'hex')

vm.runCode({
  code: code,
  gasLimit: Buffer.from('ffffffff', 'hex') 
}, function(err, results){
  console.log('returned: ' + results.return.toString('hex'));
})
```
Also more examples can be found here  
- [examples](./examples)
- [old blog post](https://wanderer.github.io/vapory/nodejs/code/2014/08/12/running-contracts-with-vm/)

# BROWSER  
To build for standalone use in the browser install `browserify` and run `npm run build`. This will give you a global variable `VapVM` to use. The standalone file will be at `./dist/vaporyjs-vm.js`

# API
  - [`new VM([StateTrie], [blockchain], [opts])`](#new-vmstatetrie-blockchain)  
  - [`VM` methods](#vm-methods)  
    - [`vm.runBlockchain([blockchain], [cb])`](#vmrunblockchainblockchain-cb)
    - [`vm.runBlock(opts, cb)`](#vmrunblockopts-cb)
    - [`vm.runTx(opts, cb)`](#vmruntxopts-cb)
    - [`vm.runCode(opts, cb)`](#vmruncodeopts-cb)
    - [`vm.generateCanonicalGenesis(cb)`](#vmgeneratecanonicalgenesiscb)
    - [`vm.generateGenesis(cb)`](#vmgenerategenesiscb)
  - [`VM` debugging hooks](#vm-debugging-hooks)
    - [`vm.onStep`](#vmonstep)

### `new VM([StateTrie], [blockchain])`
Creates a new VM object
- `StateTrie` - The [Patricia Merkle Tree](https://github.com/vaporyjs/merkle-patricia-tree) that contains the state. If no trie is given the `VM` will create an in memory trie.
- `blockchain` - an instance of the [`Blockchain`](https://github.com/vapory/vaporyjs-lib/blob/master/docs/blockchain.md) If no blockchain is given a fake blockchain will be used.
- `opts`
  - `enableHomestead` - a boolean that overrides the homestead settings based on blocknumber
  - `activatePrecompiles` - create entries in the state tree for the precompiled contracts

### `VM` methods


#### `vm.runBlockchain(blockchain, cb)`
Process a transaction.
- `blockchain` - A [blockchain](https://github.com/vapory/vaporyjs-blockchain) that to process
- `cb` - The callback. It is given an err parameter if it fails

--------------------------------------------------------

#### `vm.runBlock(opts, cb)`
Processes the `block` running all of the transactions it contains and updating the miner's account.
- `opts.block` - The [`Block`](./block.md) to process
- `opts.generate` - a `Boolean`; whether to generate the stateRoot. If false  `runBlock` will check the stateRoot of the block against the Trie
- `cb` - The callback. It is given two arguments, an `error` string containing an error that may have happened or `null`, and a `results` object with the following properties:
  - `receipts` - the receipts from the transactions in the block
  - `results` - an Array for results from the transactions in the block
--------------------------------------------------------


#### `vm.runTx(opts, cb)`
Process a transaction.
- `opts.tx` - A [`Transaction`](https://github.com/vapory/vaporyjs-tx) to run.
- `opts.block` - The block to which the `tx` belongs. If omitted a blank block will be used.
- `cb` - The callback. It is given two arguments, an `error` string containing an error that may have happened or `null`, and a `results` object with the following properties:
  - `amountSpent` - the amount of vapor used by this transaction as a `bignum`
  - `gasUsed` - the amount of gas used by the transaction
  - `vm` - contains the results from running the code, if any, as described in [`vm.runCode(params, cb)`](#vmruncodeopts-cb)

--------------------------------------------------------

#### `vm.runCode(opts, cb)`
Runs EVM code
- `opts.code` - The EVM code to run given as a `Buffer`
- `opts.data` - The input data given as a `Buffer`
- `opts.value` - The value in vapor that is being sent to `opt.address`. Defaults to `0`
- `opts.block` - The [`Block`](./block.md) the `tx` belongs to. If omitted a blank block will be used.
- `opts.gasLimit` - The gas limit for the code given as a `Buffer`
- `opts.account` - The [`Account`](./account.md) that the executing code belongs to. If omitted an empty account will be used
- `opts.address` - The address of the account that is executing this code. The address should be a `Buffer` of bytes. Defaults to `0`
- `opts.origin` - The address where the call originated from. The address should be a `Buffer` of 20bits. Defaults to `0`
- `opts.caller` - The address that ran this code. The address should be a `Buffer` of 20bits. Defaults to `0`
- `cb` - The callback. It is given two arguments, an `error` string containing an error that may have happened or `null` and a `results` object with the following properties
  - `gas` - the amount of gas left as a `bignum`
  - `gasUsed` - the amount of gas as a `bignum` the code used to run. 
  - `gasRefund` - a `Bignum` containing the amount of gas to refund from deleting storage values
  - `suicides` - an `Array` of accounts that have suicided.
  - `suicideTo` - the account that the suicide refund should go to.
  - `logs` - an `Array` of logs that the contract emitted.
  - `exception` - a `boolean`, whether or not the contract encountered an exception
  - `exceptionError` - a `String` describing the exception if there was one.
  - `return` - a `Buffer` containing the value that was returned by the contract


--------------------------------------------------------

#### `vm.stateManager.generateCanonicalGenesis(cb)`
Generates the Canonical genesis state.

--------------------------------------------------------

#### `vm.stateManager.generateGenesis(genesisData, cb)`
Generate the genesis state.
- `genesisData` - an `Object` whose keys are addresses and values are `string`s representing initial allocation of vapor.
- `cb` - The callback

```javascript
var genesisData = {
  "51ba59315b3a95761d0863b05ccc7a7f54703d99": "1606938044258990275541962092341162602522202993782792835301376",
  "e4157b34ea9615cfbde6b4fda419828124b70c78": "1606938044258990275541962092341162602522202993782792835301376"
}

vm.generateGenesis(genesisData, function(){
  console.log('generation done');
})
```

### `events`
All events are instances of [async-eventemmiter](https://www.npmjs.com/package/async-eventemitter). If an event handler has an arity of 2 the VM will pause until the callback is called

#### `step` 
The `step` event is given an `Object` and callback. The `Object` has the following properties.
- `pc` - a `Number` representing the program counter
- `opcode` - the next opcode to be ran
- `gas` - a `bignum` standing for the amount of gasLeft
- `stack` - an `Array` of `Buffers` containing the stack. 
- `storageTrie` - the storage [trie](https://github.com/vaporyjs/merkle-patricia-tree) for the account
- `account` - the [`Account`](https://github.com/vapory/vaporyjs-account) which owns the code running.
- `address` - the address of the `account`
- `depth` - the current number of calls deep the contract is
- `memory` - the memory of the VM as a `buffer`
- `cache` - The account cache. Contains all the accounts loaded from the trie. It is an instance of [functional red black tree](https://www.npmjs.com/package/functional-red-black-tree) 

#### `beforeBlock`
Emits the block that is about to be processed.

#### `afterBlock`
Emits the results of the processing a block.

#### `beforeTx`
Emits the Transaction that I about to be processed.

#### `afterTx`
Emits the result of the transaction.

# TESTING
`npm test`  
if you want to just run the Blockchain tests run
`./bin/tester -b`
if you want to just run the VM tests run
`./bin/tester -v`
if you want to just run the State tests run
`./bin/tester -s`

# Internal Structure
The VM processes state changes at many levels.

* runBlockchain
  * for every block, runBlock
* runBlock
  * for every tx, runTx
  * pay miner and uncles
* runTx
  * check sender balance
  * check sender nonce
  * runCall
  * transfer gas charges
* runCall
  * checkpoint state
  * transfer value
  * load code
  * runCode
  * materialize created contracts
  * revert or commit checkpoint
* runCode
  * iterate over code
  * run op codes
  * track gas usage
* OpFns
  * run individual op code
  * modify stack
  * modify memory
  * calculate fee

The opFns for `CREATE`, `CALL`, and `CALLCODE` call back up to `runCall`.


# LICENSE
[MPL-2.0](https://www.mozilla.org/MPL/2.0/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) 
(modification: no type change headlines) and this project adheres to 
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.3.3] - 2018-02-02

- Reworked memory expansion/access for opcodes, PR [#174](https://github.com/vaporyjs/vaporyjs-vm/pull/174) (fixes consensus bugs on
  large numbers >= 53 bit for opcodes using memory location)
- Keep stack items as bn.js instances (arithmetic performance increases), PRs [#159](https://github.com/vaporyjs/vaporyjs-vm/pull/159), [#254](https://github.com/vaporyjs/vaporyjs-vm/pull/254) and [#256](https://github.com/vaporyjs/vaporyjs-vm/pull/256)
- More consistent VM error handling, PR [#219](https://github.com/vaporyjs/vaporyjs-vm/pull/219)
- Validate stack items after operations, PR [#222](https://github.com/vaporyjs/vaporyjs-vm/pull/222)
- Updated ``vaporyjs-util`` dependency from ``4.5.0`` to ``5.1.x``, PR [#241](https://github.com/vaporyjs/vaporyjs-vm/pull/241)
- Fixed child contract deletion bug, PR [#246](https://github.com/vaporyjs/vaporyjs-vm/pull/246)
- Fixed a bug associated with direct stack usage, PR [#240](https://github.com/vaporyjs/vaporyjs-vm/pull/240)
- Fix error on large return fees, PR [#235](https://github.com/vaporyjs/vaporyjs-vm/pull/235)
- Various bug fixes

[2.3.3]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.3.2...v2.3.3

## [2.3.2] - 2017-10-29
- Better handling of ``rustbn.js`` exceptions
- Fake (default if non-provided) blockchain fixes
- Testing improvements (separate skip lists)
- Minor optimizations and bug fixes

[2.3.2]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.3.1...v2.3.2

## [2.3.1] - 2017-10-11
- ``Byzantium`` compatible
- New opcodes ``REVERT``, ``RETURNDATA`` and ``STATICCALL``
- Precompiles for curve operations and bigint mod exp
- Transaction return data in receipts
- For detailed list of changes see PR [#161](https://github.com/vaporyjs/vaporyjs-vm/pull/161) 
- For a ``Spurious Dragon``/``EIP 150`` compatible version of this library install latest version of ``2.2.x``

[2.3.1]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.2.2...v2.3.1

## [2.3.0] - Version Skipped due to faulty npm release

## [2.2.2] - 2017-09-19
- Fixed [JS number issues](https://github.com/vaporyjs/vaporyjs-vm/pull/168)
  and [certain edge cases](https://github.com/vaporyjs/vaporyjs-vm/pull/188)
- Fixed various smaller bugs and improved code consistency
- Some VM speedups
- Testing improvements
- Narrowed down dependencies for library not to break after Byzantium release

[2.2.2]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.2.1...v2.2.2

## [2.2.1] - 2017-08-04
- Fixed bug prevent the library to be used in the browser

[2.2.1]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.2.0...v2.2.1

## [2.2.0] - 2017-07-28
- ``Spurious Dragon`` & ``EIP 150`` compatible
- Detailed list of changes in pull requests [#147](https://github.com/vaporyjs/vaporyjs-vm/pull/147) and  [#143](https://github.com/vaporyjs/vaporyjs-vm/pull/143)
- Removed ``enableHomestead`` option when creating a [ new VM object](https://github.com/vaporyjs/vaporyjs-vm#new-vmstatetrie-blockchain) (pre-Homestead fork rules not supported any more)

[2.2.0]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.1.0...v2.2.0

## [2.1.0] - 2017-06-28
- Homestead compatible
- update state test runner for General State Tests

[2.1.0]: https://github.com/vaporyjs/vaporyjs-vm/compare/v2.0.1...v2.1.0

## Older releases:

- [2.0.1](https://github.com/vaporyjs/vaporyjs-vm/compare/v2.0.0...v2.0.1) - 2016-10-31
- [2.0.0](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.4.0...v2.0.0) - 2016-09-26
- [1.4.0](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.3.0...v1.4.0) - 2016-05-20
- [1.3.0](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.2.2...v1.3.0) - 2016-04-02
- [1.2.2](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.2.1...v1.2.2) - 2016-03-31
- [1.2.1](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.2.0...v1.2.1) - 2016-03-03
- [1.2.0](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.1.0...v1.2.0) - 2016-02-27
- [1.1.0](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.0.4...v1.1.0) - 2016-01-09
- [1.0.4](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.0.3...v1.0.4) - 2015-12-18
- [1.0.3](https://github.com/vaporyjs/vaporyjs-vm/compare/v1.0.0...v1.0.3) - 2015-11-27
- 1.0.0 - 2015-10-06


lib/logTable.js
const utils = require('vaporyjs-util')
const BN = utils.BN
const pow32 = new BN('010000000000000000000000000000000000000000000000000000000000000000', 16)
const pow31 = new BN('0100000000000000000000000000000000000000000000000000000000000000', 16)
const pow30 = new BN('01000000000000000000000000000000000000000000000000000000000000', 16)
const pow29 = new BN('010000000000000000000000000000000000000000000000000000000000', 16)
const pow28 = new BN('0100000000000000000000000000000000000000000000000000000000', 16)
const pow27 = new BN('01000000000000000000000000000000000000000000000000000000', 16)
const pow26 = new BN('010000000000000000000000000000000000000000000000000000', 16)
const pow25 = new BN('0100000000000000000000000000000000000000000000000000', 16)
const pow24 = new BN('01000000000000000000000000000000000000000000000000', 16)
const pow23 = new BN('010000000000000000000000000000000000000000000000', 16)
const pow22 = new BN('0100000000000000000000000000000000000000000000', 16)
const pow21 = new BN('01000000000000000000000000000000000000000000', 16)
const pow20 = new BN('010000000000000000000000000000000000000000', 16)
const pow19 = new BN('0100000000000000000000000000000000000000', 16)
const pow18 = new BN('01000000000000000000000000000000000000', 16)
const pow17 = new BN('010000000000000000000000000000000000', 16)
const pow16 = new BN('0100000000000000000000000000000000', 16)
const pow15 = new BN('01000000000000000000000000000000', 16)
const pow14 = new BN('010000000000000000000000000000', 16)
const pow13 = new BN('0100000000000000000000000000', 16)
const pow12 = new BN('01000000000000000000000000', 16)
const pow11 = new BN('010000000000000000000000', 16)
const pow10 = new BN('0100000000000000000000', 16)
const pow9 = new BN('01000000000000000000', 16)
const pow8 = new BN('010000000000000000', 16)
const pow7 = new BN('0100000000000000', 16)
const pow6 = new BN('01000000000000', 16)
const pow5 = new BN('010000000000', 16)
const pow4 = new BN('0100000000', 16)
const pow3 = new BN('01000000', 16)
const pow2 = new BN('010000', 16)
const pow1 = new BN('0100', 16)

module.exports = function (a) {
  if (a.cmp(pow1) === -1) {
    return 0
  } else if (a.cmp(pow2) === -1) {
    return 1
  } else if (a.cmp(pow3) === -1) {
    return 2
  } else if (a.cmp(pow4) === -1) {
    return 3
  } else if (a.cmp(pow5) === -1) {
    return 4
  } else if (a.cmp(pow6) === -1) {
    return 5
  } else if (a.cmp(pow7) === -1) {
    return 6
  } else if (a.cmp(pow8) === -1) {
    return 7
  } else if (a.cmp(pow9) === -1) {
    return 8
  } else if (a.cmp(pow10) === -1) {
    return 9
  } else if (a.cmp(pow11) === -1) {
    return 10
  } else if (a.cmp(pow12) === -1) {
    return 11
  } else if (a.cmp(pow13) === -1) {
    return 12
  } else if (a.cmp(pow14) === -1) {
    return 13
  } else if (a.cmp(pow15) === -1) {
    return 14
  } else if (a.cmp(pow16) === -1) {
    return 15
  } else if (a.cmp(pow17) === -1) {
    return 16
  } else if (a.cmp(pow18) === -1) {
    return 17
  } else if (a.cmp(pow19) === -1) {
    return 18
  } else if (a.cmp(pow20) === -1) {
    return 19
  } else if (a.cmp(pow21) === -1) {
    return 20
  } else if (a.cmp(pow22) === -1) {
    return 21
  } else if (a.cmp(pow23) === -1) {
    return 22
  } else if (a.cmp(pow24) === -1) {
    return 23
  } else if (a.cmp(pow25) === -1) {
    return 24
  } else if (a.cmp(pow26) === -1) {
    return 25
  } else if (a.cmp(pow27) === -1) {
    return 26
  } else if (a.cmp(pow28) === -1) {
    return 27
  } else if (a.cmp(pow29) === -1) {
    return 28
  } else if (a.cmp(pow30) === -1) {
    return 29
  } else if (a.cmp(pow31) === -1) {
    return 30
  } else if (a.cmp(pow32) === -1) {
    return 31
  } else {
    return 32
  }
}
